### ISSUE TITLE
CA2243 false positive when using ObsoleteAttribute.UrlFormat

### ISSUE THREAD

**Issue by GrabYourPitchforks** [dotnet/roslyn-analyzers#3635]
#### Analyzer package and version

(Latest checked in to dotnet/runtime.)

#### Diagnostic ID

CA2243: Attribute string literals should parse correctly
https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2243

#### Problem description

In .NET 5.0, we're extending the type `ObsoleteAttribute` by adding some new properties. See https://github.com/dotnet/runtime/issues/33089 for more info.

```cs
namespace System
{
    public partial class ObsoleteAttribute : Attribute
    {
        // New:
        public string? DiagnosticId { get; set; }
        public string? UrlFormat { get; set; }
    }
}
```

The _UrlFormat_ property is typed as _string_, and even though it has "Url" in the property name it's not intended to represent a well-formed URL. Rather, it's intended to represent a format string such as `"https://example.com/{0}/"`, where the compiler will essentially perform a `string.Format(UrlFormat, DiagnosticId)` when generating the friendly URL to display to the developer.

However, CA2243 sees the literal `'{'` and `'}'` characters in the string and deduces that the URL itself is not well-formed. This leads to false positive CA2243 hits for whoever ends up using the newly introduced extensibility features.

My recommendation is that CA2243 special-case this property and ignore it.

/cc @terrajobst

**Comment by mavasani** [comment]
Sounds good.Shoulld bail out for Obsolete attributes here: https://github.com/dotnet/roslyn-analyzers/blob/afe670e07d4648f3d2b1dc6e33dfe035966c9f63/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/AttributeStringLiteralsShouldParseCorrectly.cs#L154

**Comment by fowl2** [comment]
Rather than special casing, perhaps just ignore properties ending in "Format"?

**Comment by Evangelink** [comment]
I agree with @fowl2! If we also see the word `Format` we should not try to check that the URL is correctly formed.

**Comment by mavasani** [comment]
@Evangelink we can potentially use the WordParser and bail if any word part is "Format". Would you like to take this up?

**Comment by Evangelink** [comment]
@mavasani Happy to handle this :)

Regarding the fix, we could either:

1. Add an exception only for `UrlFormat`, re-using the mechanism in place here https://github.com/dotnet/roslyn-analyzers/blob/afe670e07d4648f3d2b1dc6e33dfe035966c9f63/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/AttributeStringLiteralsShouldParseCorrectly.cs#L256

2. We could use the `WordParser` and bail if there is `Format`.

If we go for option 2, do you want to wait for a specific request or shall I also handle the `Format` part for the `GUID` check?

**Comment by mavasani** [comment]
@Evangelink Thanks for the pointers. Both approaches seem fine to me. It seems reasonable to re-use the existing mechanism in place for validation - switch from an optional `ignoredName` to params array of `ignoredNames`, where we either ignore just the specific names or ignore names which have these as a part as identified by the WordParser.

**Comment by terrajobst** [comment]
@mavasani we should also consider special casing types derived from `Attribute` as they can't use the `Uri` type due to metadata constraints.
